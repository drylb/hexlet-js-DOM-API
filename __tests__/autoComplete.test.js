/* eslint-disable no-undef */
import fs from 'fs';
import path from 'path';
import timer from 'timer-promise';
import prettier from 'prettier';
// import { fileURLToPath } from 'url';
import { jest } from '@jest/globals';

import run from '../src/autoComplete';

/* eslint-disable no-underscore-dangle */
// const __filename = fileURLToPath(import.meta.url);
// const __dirname = path.dirname(__filename);

const options = {
  parser: 'html',
  htmlWhitespaceSensitivity: 'ignore',
  tabWidth: 4,
};

const fixuturesPath = path.join(process.cwd(), '__fixtures__');
const getTree = () => prettier.format(document.body.innerHTML, options);

const pressKey = (key, el = document.body, value = key) => {
  const e = new KeyboardEvent('input', { key, value });
  el.value = value; // eslint-disable-line
  el.dispatchEvent(e);
};

const mockResponse = (status, statusText, response) => new window.Response(response, {
  status,
  statusText,
  headers: {
    'Content-type': 'application/json',
  },
});

const mockFetch = (data) => {
  window.fetch = jest.fn().mockImplementation(() => (
    Promise.resolve(mockResponse(200, null, JSON.stringify(data)))));
};

beforeEach(async () => {
  const initHtml = fs.readFileSync(path.join(fixuturesPath, 'autoComplete.html')).toString();
  document.documentElement.innerHTML = initHtml;
  run();
});

test('application', async () => {
  const input = document.querySelector('[data-autocomplete="/capitals.json"]');
  input.focus();

  expect(getTree()).toMatchSnapshot();

  mockFetch([]);
  pressKey('x', input);
  await timer.start(100);
  expect(getTree()).toMatchSnapshot();

  mockFetch([]);
  pressKey('backspace', input, '');
  await timer.start(100);
  expect(getTree()).toMatchSnapshot();

  mockFetch(['mariehamn', 'manama', 'minsk', 'moroni']);
  pressKey('m', input);
  await timer.start(100);
  expect(getTree()).toMatchSnapshot();

  mockFetch(['moroni', 'monrovia', 'monaco', 'moscow']);
  pressKey('o', input, 'mo');
  await timer.start(100);
  expect(getTree()).toMatchSnapshot();

  mockFetch(['moscow']);
  pressKey('s', input, 'mos');
  await timer.start(100);
  expect(getTree()).toMatchSnapshot();
});

test('application 2', async () => {
  const input = document.querySelector('[data-autocomplete="/countries.json"]');
  input.focus();

  mockFetch([]);
  pressKey('r', input);
  await timer.start(100);
  expect(getTree()).toMatchSnapshot();

  mockFetch(['russia']);
  pressKey('u', input, 'ru');
  await timer.start(100);
  expect(getTree()).toMatchSnapshot();

  mockFetch([]);
  pressKey('c', input, 'ruc');
  await timer.start(100);
  expect(getTree()).toMatchSnapshot();
});
