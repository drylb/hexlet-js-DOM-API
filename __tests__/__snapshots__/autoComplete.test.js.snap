// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application 1`] = `
"<div class=\\"container\\">
    <div class=\\"row\\">
        <div class=\\"col-6\\">
            <form class=\\"form-inline\\">
                <div class=\\"form-group\\">
                    <label>Capital</label>
                    <input
                        class=\\"ml-2 form-control\\"
                        type=\\"text\\"
                        data-autocomplete-name=\\"capital\\"
                        data-autocomplete=\\"/capitals.json\\"
                    />
                </div>
            </form>
        </div>
        <div class=\\"col-6\\">
            <ul data-autocomplete-name=\\"capital\\">
                <li>Nothing</li>
            </ul>
        </div>
    </div>
</div>
<div class=\\"container\\">
    <div class=\\"row\\">
        <div class=\\"col-6\\">
            <form class=\\"form-inline\\">
                <div class=\\"form-group\\">
                    <label>Country</label>
                    <input
                        class=\\"ml-2 form-control\\"
                        type=\\"text\\"
                        data-autocomplete-name=\\"country\\"
                        data-autocomplete=\\"/countries.json\\"
                    />
                </div>
            </form>
        </div>
        <div class=\\"col-6\\">
            <ul data-autocomplete-name=\\"country\\">
                <li>Nothing</li>
            </ul>
        </div>
    </div>
</div>
"
`;

exports[`application 2`] = `
"<div class=\\"container\\">
    <div class=\\"row\\">
        <div class=\\"col-6\\">
            <form class=\\"form-inline\\">
                <div class=\\"form-group\\">
                    <label>Capital</label>
                    <input
                        class=\\"ml-2 form-control\\"
                        type=\\"text\\"
                        data-autocomplete-name=\\"capital\\"
                        data-autocomplete=\\"/capitals.json\\"
                    />
                </div>
            </form>
        </div>
        <div class=\\"col-6\\">
            <ul data-autocomplete-name=\\"capital\\">
                <li>Nothing</li>
            </ul>
        </div>
    </div>
</div>
<div class=\\"container\\">
    <div class=\\"row\\">
        <div class=\\"col-6\\">
            <form class=\\"form-inline\\">
                <div class=\\"form-group\\">
                    <label>Country</label>
                    <input
                        class=\\"ml-2 form-control\\"
                        type=\\"text\\"
                        data-autocomplete-name=\\"country\\"
                        data-autocomplete=\\"/countries.json\\"
                    />
                </div>
            </form>
        </div>
        <div class=\\"col-6\\">
            <ul data-autocomplete-name=\\"country\\">
                <li>Nothing</li>
            </ul>
        </div>
    </div>
</div>
"
`;

exports[`application 2 1`] = `
"<div class=\\"container\\">
    <div class=\\"row\\">
        <div class=\\"col-6\\">
            <form class=\\"form-inline\\">
                <div class=\\"form-group\\">
                    <label>Capital</label>
                    <input
                        class=\\"ml-2 form-control\\"
                        type=\\"text\\"
                        data-autocomplete-name=\\"capital\\"
                        data-autocomplete=\\"/capitals.json\\"
                    />
                </div>
            </form>
        </div>
        <div class=\\"col-6\\">
            <ul data-autocomplete-name=\\"capital\\">
                <li>Nothing</li>
            </ul>
        </div>
    </div>
</div>
<div class=\\"container\\">
    <div class=\\"row\\">
        <div class=\\"col-6\\">
            <form class=\\"form-inline\\">
                <div class=\\"form-group\\">
                    <label>Country</label>
                    <input
                        class=\\"ml-2 form-control\\"
                        type=\\"text\\"
                        data-autocomplete-name=\\"country\\"
                        data-autocomplete=\\"/countries.json\\"
                    />
                </div>
            </form>
        </div>
        <div class=\\"col-6\\">
            <ul data-autocomplete-name=\\"country\\">
                <li>Nothing</li>
            </ul>
        </div>
    </div>
</div>
"
`;

exports[`application 2 2`] = `
"<div class=\\"container\\">
    <div class=\\"row\\">
        <div class=\\"col-6\\">
            <form class=\\"form-inline\\">
                <div class=\\"form-group\\">
                    <label>Capital</label>
                    <input
                        class=\\"ml-2 form-control\\"
                        type=\\"text\\"
                        data-autocomplete-name=\\"capital\\"
                        data-autocomplete=\\"/capitals.json\\"
                    />
                </div>
            </form>
        </div>
        <div class=\\"col-6\\">
            <ul data-autocomplete-name=\\"capital\\">
                <li>Nothing</li>
            </ul>
        </div>
    </div>
</div>
<div class=\\"container\\">
    <div class=\\"row\\">
        <div class=\\"col-6\\">
            <form class=\\"form-inline\\">
                <div class=\\"form-group\\">
                    <label>Country</label>
                    <input
                        class=\\"ml-2 form-control\\"
                        type=\\"text\\"
                        data-autocomplete-name=\\"country\\"
                        data-autocomplete=\\"/countries.json\\"
                    />
                </div>
            </form>
        </div>
        <div class=\\"col-6\\">
            <ul data-autocomplete-name=\\"country\\">
                <li>russia</li>
            </ul>
        </div>
    </div>
</div>
"
`;

exports[`application 2 3`] = `
"<div class=\\"container\\">
    <div class=\\"row\\">
        <div class=\\"col-6\\">
            <form class=\\"form-inline\\">
                <div class=\\"form-group\\">
                    <label>Capital</label>
                    <input
                        class=\\"ml-2 form-control\\"
                        type=\\"text\\"
                        data-autocomplete-name=\\"capital\\"
                        data-autocomplete=\\"/capitals.json\\"
                    />
                </div>
            </form>
        </div>
        <div class=\\"col-6\\">
            <ul data-autocomplete-name=\\"capital\\">
                <li>Nothing</li>
            </ul>
        </div>
    </div>
</div>
<div class=\\"container\\">
    <div class=\\"row\\">
        <div class=\\"col-6\\">
            <form class=\\"form-inline\\">
                <div class=\\"form-group\\">
                    <label>Country</label>
                    <input
                        class=\\"ml-2 form-control\\"
                        type=\\"text\\"
                        data-autocomplete-name=\\"country\\"
                        data-autocomplete=\\"/countries.json\\"
                    />
                </div>
            </form>
        </div>
        <div class=\\"col-6\\">
            <ul data-autocomplete-name=\\"country\\">
                <li>Nothing</li>
            </ul>
        </div>
    </div>
</div>
"
`;

exports[`application 3`] = `
"<div class=\\"container\\">
    <div class=\\"row\\">
        <div class=\\"col-6\\">
            <form class=\\"form-inline\\">
                <div class=\\"form-group\\">
                    <label>Capital</label>
                    <input
                        class=\\"ml-2 form-control\\"
                        type=\\"text\\"
                        data-autocomplete-name=\\"capital\\"
                        data-autocomplete=\\"/capitals.json\\"
                    />
                </div>
            </form>
        </div>
        <div class=\\"col-6\\">
            <ul data-autocomplete-name=\\"capital\\">
                <li>Nothing</li>
            </ul>
        </div>
    </div>
</div>
<div class=\\"container\\">
    <div class=\\"row\\">
        <div class=\\"col-6\\">
            <form class=\\"form-inline\\">
                <div class=\\"form-group\\">
                    <label>Country</label>
                    <input
                        class=\\"ml-2 form-control\\"
                        type=\\"text\\"
                        data-autocomplete-name=\\"country\\"
                        data-autocomplete=\\"/countries.json\\"
                    />
                </div>
            </form>
        </div>
        <div class=\\"col-6\\">
            <ul data-autocomplete-name=\\"country\\">
                <li>Nothing</li>
            </ul>
        </div>
    </div>
</div>
"
`;

exports[`application 4`] = `
"<div class=\\"container\\">
    <div class=\\"row\\">
        <div class=\\"col-6\\">
            <form class=\\"form-inline\\">
                <div class=\\"form-group\\">
                    <label>Capital</label>
                    <input
                        class=\\"ml-2 form-control\\"
                        type=\\"text\\"
                        data-autocomplete-name=\\"capital\\"
                        data-autocomplete=\\"/capitals.json\\"
                    />
                </div>
            </form>
        </div>
        <div class=\\"col-6\\">
            <ul data-autocomplete-name=\\"capital\\">
                <li>mariehamn</li>
                <li>manama</li>
                <li>minsk</li>
                <li>moroni</li>
            </ul>
        </div>
    </div>
</div>
<div class=\\"container\\">
    <div class=\\"row\\">
        <div class=\\"col-6\\">
            <form class=\\"form-inline\\">
                <div class=\\"form-group\\">
                    <label>Country</label>
                    <input
                        class=\\"ml-2 form-control\\"
                        type=\\"text\\"
                        data-autocomplete-name=\\"country\\"
                        data-autocomplete=\\"/countries.json\\"
                    />
                </div>
            </form>
        </div>
        <div class=\\"col-6\\">
            <ul data-autocomplete-name=\\"country\\">
                <li>Nothing</li>
            </ul>
        </div>
    </div>
</div>
"
`;

exports[`application 5`] = `
"<div class=\\"container\\">
    <div class=\\"row\\">
        <div class=\\"col-6\\">
            <form class=\\"form-inline\\">
                <div class=\\"form-group\\">
                    <label>Capital</label>
                    <input
                        class=\\"ml-2 form-control\\"
                        type=\\"text\\"
                        data-autocomplete-name=\\"capital\\"
                        data-autocomplete=\\"/capitals.json\\"
                    />
                </div>
            </form>
        </div>
        <div class=\\"col-6\\">
            <ul data-autocomplete-name=\\"capital\\">
                <li>moroni</li>
                <li>monrovia</li>
                <li>monaco</li>
                <li>moscow</li>
            </ul>
        </div>
    </div>
</div>
<div class=\\"container\\">
    <div class=\\"row\\">
        <div class=\\"col-6\\">
            <form class=\\"form-inline\\">
                <div class=\\"form-group\\">
                    <label>Country</label>
                    <input
                        class=\\"ml-2 form-control\\"
                        type=\\"text\\"
                        data-autocomplete-name=\\"country\\"
                        data-autocomplete=\\"/countries.json\\"
                    />
                </div>
            </form>
        </div>
        <div class=\\"col-6\\">
            <ul data-autocomplete-name=\\"country\\">
                <li>Nothing</li>
            </ul>
        </div>
    </div>
</div>
"
`;

exports[`application 6`] = `
"<div class=\\"container\\">
    <div class=\\"row\\">
        <div class=\\"col-6\\">
            <form class=\\"form-inline\\">
                <div class=\\"form-group\\">
                    <label>Capital</label>
                    <input
                        class=\\"ml-2 form-control\\"
                        type=\\"text\\"
                        data-autocomplete-name=\\"capital\\"
                        data-autocomplete=\\"/capitals.json\\"
                    />
                </div>
            </form>
        </div>
        <div class=\\"col-6\\">
            <ul data-autocomplete-name=\\"capital\\">
                <li>moscow</li>
            </ul>
        </div>
    </div>
</div>
<div class=\\"container\\">
    <div class=\\"row\\">
        <div class=\\"col-6\\">
            <form class=\\"form-inline\\">
                <div class=\\"form-group\\">
                    <label>Country</label>
                    <input
                        class=\\"ml-2 form-control\\"
                        type=\\"text\\"
                        data-autocomplete-name=\\"country\\"
                        data-autocomplete=\\"/countries.json\\"
                    />
                </div>
            </form>
        </div>
        <div class=\\"col-6\\">
            <ul data-autocomplete-name=\\"country\\">
                <li>Nothing</li>
            </ul>
        </div>
    </div>
</div>
"
`;
